### PR TITLE
Improve source file remapping

### PR DESCRIPTION
**Pull request by jairbubbles**
When running tests on Linux I was getting the following crash:

```
10:57:45   Error Message:
10:57:45    System.ArgumentNullException : Value cannot be null. (Parameter 'directory')
10:57:45   Stack Trace:
10:57:45      at Guard.AgainstEmpty(String value, String argumentName) in /_/src/Verify/Guard.cs:line 108
10:57:45    at Guard.BadDirectoryName(String name, String argumentName) in /_/src/Verify/Guard.cs:line 59
10:57:45    at VerifyTests.PathInfo..ctor(String directory, String typeName, String methodName) in /_/src/Verify/DerivePaths/PathInfo.cs:line 15
10:57:45    at VerifyTests.PathInfo.DeriveDefault(String sourceFile, String projectDirectory, Type type, MethodInfo method) in /_/src/Verify/DerivePaths/PathInfo.cs:line 30
10:57:45    at VerifyNUnit.Verifier.GetPathInfo(String sourceFile, Type type, MethodInfo method) in /_/src/Verify.NUnit/DerivePaths/VerifierSettings.cs:line 10
10:57:45    at VerifyNUnit.Verifier.BuildVerifier(String sourceFile, VerifySettings settings, Boolean useUniqueDirectory) in /_/src/Verify.NUnit/Verifier.cs:line 51
```

After investigation, it appears that:
- My project is built on Windows
- The resulting assemblies are passed to another Linux machine to run tests
- In that scenario, source file path retrieved through `[callerFilePath]` attribute is a Windows path and `Path.GetDirectoryName` will return `null` with such a path on Linux.

I then stumbled on `GetMappedBuildPath` (introduced in https://github.com/VerifyTests/Verify/pull/729) but it's not detecting my case as I'm not running in a WSL context.

After some discussions with @JanKrivanek, we ended up:
- Stop using `Path.GetDirectoryName` as it returns null with a Windows path on Linux
- The method is now called `IoHelpers.ResolveDirectoryFromSourceFile` to be more honest about what it does
- Improve the remapping algorithm to cover more use cases (new approach is described [here](https://github.com/VerifyTests/Verify/pull/775#discussion_r1098330840))
- Make the code testable (by introducing a new interface `IEnvironment`) and add some tests to cover the old and new use cases

### CODE REVIEW

**Comment by SimonCropp** [comment]
@JanKrivanek thoughts

**Comment by JanKrivanek** [comment]
I'll need to have a closer look and thought.. but the very first idea is that it'll crash anyway down the road :-/
Basically `Verify` relies on build-time paths info - that is completely lost once the build outputs are copied to another machine. Normalizning the paths will unblock the `Path.GetDirectoryName`, but the results will be a nonexistent path. I'm cannot now recall what the caller info is exactly used for - is it just naming? Then it might be proably overriden via custom naming info (`UseTypeName`, `UseMethodName`).

**tl;dr;** "What is your scenarion?" :-) We might have similar in https://github.com/dotnet/templating/blob/main/tools/Microsoft.TemplateEngine.Authoring.TemplateVerifier/VerificationEngine.cs

File: src/Verify/IoHelpers.cs [hunk 1]
```diff
@@ -201,8 +201,15 @@ public static string GetRelativePath(string directory, string file)
...
 
-    public static string? GetDirectoryName(string? path) =>
-        Path.GetDirectoryName(GetMappedBuildPath(path));
+    public static string? GetDirectoryName(string? path)
+    {
+        var mappedPath = GetMappedBuildPath(path);
+        // Path remapping might not always be working (ex: you've built your project on Windows and you're running tests
+        // on the same assemblies but on another Linux machine). Always use / so that Path.GetDirectoryName does not
+        // return null on Unix systems.
+        mappedPath = mappedPath?.Replace("\\", "/");
```
Comment (JanKrivanek) [inline src/Verify/IoHelpers.cs:210]: If we want to handle x-platfrom, then we should probably handle both directions (so e,g, as well a unix built binaries called on win).

so something like

```C#
static readonly char[] separators =
{
    '\\',
    '/'
};

// ...

if (!string.IsNullOrEmpty(mappedPath) && !mappedPath.Contains(Path.DirectorySeparatorChar))
{
  char contraSeparator = separators.First(c => !c.Equals(Path.DirectorySeparatorChar));
  mappedPath = mappedPath.Replace(contraSeparator, Path.DirectorySeparatorChar);
}
```

**Review by JanKrivanek — CHANGES_REQUESTED** [review]
Updating [`IoHelpers.GetDirectoryName`](https://github.com/VerifyTests/Verify/blob/main/src/Verify/IoHelpers.cs#L204) to work properly on x-platform paths is probably harmless and not a bad idea. So once it's really x-platform - I'm good with the change.

It is possible that you might as well be able to workaround your situation via [injecting custom `DerivePathInfo` callback](https://github.com/VerifyTests/Verify/blob/main/src/Verify.NUnit/DerivePaths/VerifierSettings.cs#L20) and [explicitly passing caller location into `Verify` call(s)](https://github.com/VerifyTests/Verify/blob/main/src/Verify.Xunit/Verifier.cs#L56) - the path that you'd obtain in the env that is running the tests (not the build time path). 
As those are the only codepaths that end up calling into [`IoHelpers.GetDirectoryName`](https://github.com/VerifyTests/Verify/blob/main/src/Verify/IoHelpers.cs#L204)

**Comment by jairbubbles** [comment]
I feel like bypassing `Path.GetDirectoryName` totally is probably the best as it's meant to be used on a local path. In this case, the path can have different separators so having its own logic would be better and will have a consistent behavior across platforms.

I would rename it to `IoHelpers.GetDirectoryNameFromSourceFile` as the input would always be a file path.

File: src/Verify/IoHelpers.cs [hunk 2]
```diff
@@ -201,8 +202,25 @@ public static string GetRelativePath(string directory, string file)
...
 
-    public static string? GetDirectoryName(string? path) =>
-        Path.GetDirectoryName(GetMappedBuildPath(path));
+    /// <summary>
+    /// Get directory path from a given file path
+    /// </summary>
+    /// <remarks>
+    /// The method accepts input paths mixing / and \ separators
+    /// </remarks>
+    public static string GetDirectoryFromSourceFile(string? sourceFile)
```
Comment (JanKrivanek) [inline src/Verify/IoHelpers.cs:211]: There are cases where this function is not on par with the `Path.GetDirectoryName` - is that ok?

File: src/Verify.Tests/IoHelpersTests.cs [hunk 3]
```diff
@@ -0,0 +1,24 @@
...
+    [InlineData(@"C:/test/sub_folder/my_file.cs", "C:/test/sub_folder")]
+    [InlineData(@"/mnt/d/test/my_file.cs", "/mnt/d/test")]
+    [InlineData(@"/mnt/MyFile.cs", "/mnt")]
+    [InlineData(@"C:\Program Files/MyFile.cs", @"C:\Program Files")]
+    [InlineData(@"C:\MyFile.cs", @"C:")]
+    [InlineData(@"MyFile.cs", "")]
+    [InlineData(null, "")]
+    public void GetDirectoryNameFromSourceFileTests(string sourceFile, string expectedDirectory)
+    {
+        if (RuntimeInformation.IsOSPlatform(OSPlatform.Windows) && expectedDirectory != "C:")
```
[the remark this answers is not in the document]
Comment (jairbubbles) [reply src/Verify.Tests/IoHelpersTests.cs:17]: GetDirectoryName returns C:\

File: src/Verify/IoHelpers.cs (hunk 2 above)
Comment (jairbubbles) [reply src/Verify/IoHelpers.cs:211]: *Sorry for the really short answer, I was on my phone.*

My point is to make it work for what we need and only for that. If it's called `Path.GetDirectoryName` people could think it has the same behavior.

As for the differences with `Path.GetDirectoryName`   I feel like it's not much a problem but I don't really know the lib.

It will be a matter of differences of `/` and `\`. 

The .NET Framework and .NET Core implementation are quite different:
https://source.dot.net/#System.Private.CoreLib/src/libraries/System.Private.CoreLib/src/System/IO/Path.cs,132
https://referencesource.microsoft.com/#mscorlib/system/io/path.cs,156

The .NET Core implementation is really simple, we could copy the implementation and adapt it to support both separators. It would avoid any regression with the current behavior.

Comment (jairbubbles) [reply src/Verify/IoHelpers.cs:211]: Humm in fact the Windows implementation has a lot of edge cases: [Windows](https://github.com/dotnet/runtime/blob/master/src/libraries/Common/src/System/IO/PathInternal.Windows.cs) vs [Unix](https://github.com/dotnet/runtime/blob/main/src/libraries/Common/src/System/IO/PathInternal.Unix.cs)

Comment (JanKrivanek) [reply src/Verify/IoHelpers.cs:211]: I'm slightli split - probably the best solution would be to mark it `internal` (if possible) - so that it's really used in that specific case - for which the implementation is good enough.

But I do not want to be nitpicking :-)

Overall it looks good and if it solves the case of non-docker, non-WSL cross platform run - then it has it's value and I'm good with it.

**Review by JanKrivanek — APPROVED** [review]

File: src/Verify/IoHelpers.cs (hunk 2 above)
Comment (jairbubbles) [reply src/Verify/IoHelpers.cs:211]: After thinking more about it I feel like your initial feeling is true:

> but the very first idea is that it'll crash anyway down the road :-/

Is the directory produced will be used to open the verified .txt file and compare the result? If so, we would need to improve `GetMappedBuildPath` to handle non WSL cases.

Comment (jairbubbles) [reply src/Verify/IoHelpers.cs]: @JanKrivanek I updated `VirtualizedRunHelper` to handle more cases (not just WSL/Docker) and also make it unit test friendly. Let me know what you think.

File: src/Verify/VirtualizedRunHelper.cs [hunk 4]
```diff
@@ -12,82 +14,77 @@ class VirtualizedRunHelper
...
+        : this(GetOriginalCodeBaseRoot(userAssembly),
+            Environment.CurrentDirectory,
+            Path.DirectorySeparatorChar,
+            path => File.Exists(path) || Directory.Exists(path))
+    {
+    }
+
+    private static string GetOriginalCodeBaseRoot(Assembly userAssembly) =>
+        AttributeReader.TryGetSolutionDirectory(userAssembly, false, out var solutionDir)
+            ? solutionDir : AttributeReader.GetProjectDirectory(userAssembly);
```
Comment (JanKrivanek) [inline src/Verify/VirtualizedRunHelper.cs:26]: Careful - the https://github.com/VerifyTests/Verify/pull/776 stack overflow fix is not contained here

File: src/Verify/VirtualizedRunHelper.cs [hunk 5]
```diff
@@ -12,82 +14,77 @@ class VirtualizedRunHelper
...
+
+    internal VirtualizedRunHelper(string originalCodeBaseRoot, string currentDir, char runtimeSeparator, Func<string, bool> pathExists)
```
Comment (JanKrivanek) [inline src/Verify/VirtualizedRunHelper.cs:28]: `runtimeSeparator` is a weird naming (not clear what it is by just looking on the name) - I'd keep it `directorySeparatorChar`

File: src/Verify.Tests/VirtualizedRunHelperTests.cs [hunk 6]
```diff
@@ -0,0 +1,32 @@
+﻿public class VirtualizedRunHelperTests
```
Comment (JanKrivanek) [inline src/Verify.Tests/VirtualizedRunHelperTests.cs:1]: Please add a unit test for testrun on docker as well. - it mounts to `/mnt/approot/<path>`, where `<path>` is your path relative to the solution.
E.g. with the current structure:
```
C:\my-src\contoso\proj-x
                   |-- solution.sln
                   |-- src\...
                   |-- build-outputs
                          |--my-test-binary.dll
```

When running tests from the `my-test-binary.dll` the `Environment.CurrentDirectory` would return `/mnt/approot/build-outputs`

Current code seems to handle this just fine - but it's good to be future proof (e.g. if someone would try to find drive root and map paths from there)

Comment (JanKrivanek) [reply src/Verify.Tests/VirtualizedRunHelperTests.cs:1]: I'd as well recommend adding a test for case where `originalCodeBaseRoot` and `currentDir` are more distant then just single direcotry appart.
This would explicitly test a case of more nested project structure (e.g. running tests from proj `C:\my-src\contoso\proj-x\src\data-layer\utils\proj-b.fsproj` while the solution directory is `C:\my-src\contoso\proj-x`)

File: src/Verify/VirtualizedRunHelper.cs [hunk 7]
```diff
@@ -1,9 +1,11 @@
 class VirtualizedRunHelper
 {
     // e.g. WSL or docker run (https://github.com/VerifyTests/Verify#unit-testing-inside-virtualized-environment)
-    bool appearsToBeLocalVirtualizedRun;
-    string originalCodeBaseRootAbsolute = string.Empty;
-    string mappedCodeBaseRootAbsolute = string.Empty;
+    internal bool AppearsToBeLocalVirtualizedRun { get; }
+
+    string _originalCodeBaseRootAbsolute = string.Empty;
+    string _mappedCodeBaseRootAbsolute = string.Empty;
```
Comment (JanKrivanek) [inline src/Verify/VirtualizedRunHelper.cs:7]: while the fields prefix doesn't seem to be explicitly disallowed (https://github.com/VerifyTests/Verify/blob/main/src/.editorconfig#L22-L25), it doesn't seem to be used convention in this project.

**Review by JanKrivanek — CHANGES_REQUESTED** [review]
I love the indirection allowing the unit-testing!

I left couple nit-picking comments.. but my main concern is that this is not based on latest main - so it's suspectible to #774

File: src/Verify/VirtualizedRunHelper.cs (hunk 7 above)
Comment (jairbubbles) [reply src/Verify/VirtualizedRunHelper.cs:7]: Yes I hesitated, will use `this.` then.

File: src/Verify.Tests/VirtualizedRunHelperTests.cs (hunk 6 above)
Comment (jairbubbles) [reply src/Verify.Tests/VirtualizedRunHelperTests.cs:1]: I added the test (but disabled) as I'm not sure how it's supposed to work as there no common part between `/mnt/approot/build-outputs` and `C:\my-src\contoso\proj-x`

Comment (jairbubbles) [reply src/Verify.Tests/VirtualizedRunHelperTests.cs:1]: I might be wrong but I feel like we should on a given assembly test first "solution path" and then fallback on "project path".

Comment (JanKrivanek) [reply src/Verify.Tests/VirtualizedRunHelperTests.cs:1]: Good point! - this case (docker run for a build where build outputs are configured to go to the sibling folder of a .sln, while both (.sln and output folder) being directly in the 'application root') was not handled in my original code (it would refuse to map the path and continue with it as is).

While it feels outside of the scope of your original intent - would you want to fix that case as well?
Basically:
 - (already done by the code) First we try to find mapping by finding a 'cross section' of the buildtime and runtime path (e.g. `C:\build\project_dir`, `/mnt/c/build/project_dir/subdir`, have the `build/project_dir` cross section)
 - (missing handling) - we are not able to find the cross section, so we defer the initialization until the first call to `GetMappedBuildPath` requesting to map the path (`pathToMap` in following text). Then 
    - Get the relative build-time path of the `pathToMap` (removing the solution-dir prefix from `pathToMap`)
    - Iterating through current runtime path and trying to append the relative part of `pathToMap` and checking if path exists. Once it exists - we are initialized. If we iterated to the root, we are not able to perform maping for further requests.
    - Sample: current runtime dir: `/mnt/approot/build-outputs`, solution root: `C:\my-src\contoso\proj-x`, `pathToMap`: `C:\my-src\contoso\proj-x\src\data-layer\utils\proj-b.fsproj`.
      - relative requested path:  `src\data-layer\utils\proj-b.fsproj`
      - first attempted path check: `/mnt/approot/build-outputs/src/data-layer/utils/proj-b.fsproj` - doesn't exist
      - second attempted path check: `/mnt/approot/src/data-layer/utils/proj-b.fsproj` - exists, so we have the build time root `C:\my-src\contoso\proj-x\` and runtime map root `/mnt/approot/`


> I might be wrong but I feel like we should on a given assembly test first "solution path" and then fallback on "project path".

Alternatively we can try to use the "project path" (`AttributeReader.TryGetProjectDirectory`) in lieu of the `pathToMap` (from first call to `GetMappedBuildPath`) - and so being able to resolve the second case fully in the constructor, without waiting for the first call to `GetMappedBuildPath`. I guess that's what you meant?
If yes - than it sounds great - I'd just probably leave the door open for fallback from the first call to `GetMappedBuildPath` in case the `AttributeReader.TryGetProjectDirectory` would not return anyhting, or that it would return same path as `AttributeReader.TryGetSolutionDirectory`



Not sure if it sounds reasonable or if you had differnet solution in mind?

**Comment by JanKrivanek** [comment]
Overall looks good to me.
Albeit beyond origninal scope - it would be nice to include handling of the 'build outputs lives next to the solution dir' case as well

File: src/Verify.Tests/VirtualizedRunHelperTests.cs (hunk 6 above)
Comment (jairbubbles) [reply src/Verify.Tests/VirtualizedRunHelperTests.cs:1]: Yes "project dir" could provide some info to get the common part, but it won't be enough in the case "build output next to .sln".

I'm not sure I will find more time to work on this, is it ok if we merge this PR first?

**Comment by JanKrivanek** [comment]
@jairbubbles Can you please add me write permissions to your fork?
I've created a change that handles the case of execution folder being sibling to source folder.

Thanks

**Comment by jairbubbles** [comment]
Sure! You should have received an invite.

**Comment by jairbubbles** [comment]
Looking good, I updated the PR title/description.

**Comment by jairbubbles** [comment]
@SimonCropp If you're willing to merge, I believe you have an option to coauthor when squashing, @JanKrivanek made most of the work so he should get credit.

**Comment by SimonCropp** [comment]
@jairbubbles @JanKrivanek thanks for all the effort on this one.

File: src/Verify.Tests/IoHelpersTests.cs [hunk 8]
```diff
@@ -0,0 +1,23 @@
...
+public class IoHelpersTests
+{
+    [Theory]
+    [InlineData(@"C:\test\my_file.cs", @"C:\test")]
+    [InlineData(@"C:/test/sub_folder/my_file.cs", "C:/test/sub_folder")]
+    [InlineData(@"/mnt/d/test/my_file.cs", "/mnt/d/test")]
+    [InlineData(@"/mnt/MyFile.cs", "/mnt")]
+    [InlineData(@"C:\Program Files/MyFile.cs", @"C:\Program Files")]
+    [InlineData(@"C:\MyFile.cs", @"C:")]
+    [InlineData(@"MyFile.cs", "")]
```
Comment (SimonCropp) [inline src/Verify.Tests/IoHelpersTests.cs:12]: why this test? in what scenario will a non-full path be passed in?

**Comment by SimonCropp** [comment]
this is deployed. can u smoke test

**Comment by JanKrivanek** [comment]
> this is deployed. can u smoke test

looks good for basic scenarios

**Comment by jairbubbles** [comment]
On my side my CI on Linux / macOS is now fixed.
